semgrep-core/matching/Rules_filter.ml

```diff
@@ -0,0 +1,155 @@
+ * than Re.
+ *
+ * todo:
+ *  - we could even avoid parsing the file by using a lazy AST
+ *    in Semgrep_generic.check.
+ *)
+
+(*****************************************************************************)
+(* Helpers  *)
+(*****************************************************************************)
+
+(* Those functions using Re instead of Str are actually slower on
+ * Zulip.
+let regexp_matching_str s =
+  Re.str s
+let compile_regexp t =
+  Common.profile_code "Re.compile_regexp" (fun () -> Re.compile t)
+let run_regexp re str =
+  Common.profile_code "Re.run_regexp" (fun () -> Re.execp re str)
+*)
+
+let regexp_matching_str s =
+  Str.regexp_string s
+let compile_regexp t =
+  Common.profile_code "Re.compile_regexp" (fun () -> t)
+let run_regexp re str =
+  Common.profile_code "Re.run_regexp" (fun () ->
+   (* bugfix:
+    * this does not work!:  Str.string_match re str 0
+    * because you need to add ".*" in front to make it work,
+    * (but then you can not use regexp_string above)
+    * => use Str.search_forward instead.
+    *)
+   try
+     Str.search_forward re str 0 |> ignore; true
+   with Not_found -> false
+  )
+
+let reserved_id lang str =
+  Metavars_generic.is_metavar_name str ||
+  (* in JS field names can be regexps *)
+  (lang = Lang.Javascript && Matching_generic.is_regexp_string str) ||
+  (* ugly hack that we then need to handle also here *)
+  str = AST_generic.special_multivardef_pattern ||
+  (* ugly: because ast_js_build introduce some extra "!default" ids *)
+  (lang = Lang.Javascript && str = Ast_js.default_entity) ||
+  (* parser_js.mly inserts some implicit this *)
+  (lang = Lang.Java && str = "this")
+
+let reserved_str str =
+  str = "..." ||
+  Matching_generic.is_regexp_string str
+
+(*****************************************************************************)
+(* ID extractor  *)
+(*****************************************************************************)
+let extract_specific_strings lang any =
+  let res = ref [] in
+  let visitor = V.mk_visitor {V.default_visitor with
+     V.kident = (fun (_k, _) (str, _tok) ->
+       if not (reserved_id lang str)
+       then Common.push str res
+     );
+     V.kexpr = (fun (k, _) x ->
+       (match x with
+       (* less: we could extract strings for the other literals too?
+        * atoms, chars, even int?
+        *)
+       | L (String (str, _tok)) ->
+         if not (reserved_str str)
+         then Common.push str res
+       (* do not recurse there, the type does not have to be in the source *)
+       | TypedMetavar _ ->
+          ()
+       | _ -> k x
+       );
+     );
+  } in
+  visitor any;
+  !res
+
+(*****************************************************************************)
+(* Entry point *)
+(*****************************************************************************)
+
+let filter_rules_relevant_to_file_using_regexp2 rules lang file =
+  let str = Common.read_file file in
```

Comment:
Have you looked at the implications of running this on, e.g., a large minified js file? It looks like we need to load the entire file into a single block of memory here (not sure how the internals of ocaml Str work, but I assume it wraps a contiguous character array).

It might be worth wrapping this in a max memory guard.

Presumably a streaming solution is even better, but probably an annoying implementation.

Reply:
if we're slow with read_file, it will be far slower with parsing it and matching it.

Reply:
Isn't parsing guarded by max_memory though? Or does that also apply to this load?

Reply:
The whole parsing and matching is guarded by max_memory. But again, if we exceed memory with read_file, there's no way we can generate the AST for this file, which uses far more memory.